File: src/main/java/Proxy.java

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.channels.*;
import java.util.Iterator;
import java.util.Set;


public class Proxy {
    private ProxyOperations proxyOperations = new ProxyOperations();

    void run(int port) {
        Selector selector;
        ServerSocket socket;
        try {
            ServerSocketChannel socketChanel = ServerSocketChannel.open();
            socketChanel.configureBlocking(false);
            socket = socketChanel.socket();
            socket.bind(new InetSocketAddress(port));
            selector = Selector.open();
            socketChanel.register(selector, SelectionKey.OP_ACCEPT);
            DatagramChannel dnsChannel = proxyOperations.createDatagramChannel();
            if (dnsChannel != null)
                dnsChannel.register(selector, SelectionKey.OP_READ);

        } catch (IOException e) {
            e.printStackTrace();
            return;
        }

        while (true) {
            int ready = 0;
            try {
                ready = selector.select();
            } catch (IOException e) {
                e.printStackTrace();
                return;
            }
            if (ready == 0)
                continue;
            Set<SelectionKey> keys = selector.selectedKeys();
            Iterator<SelectionKey> cur = keys.iterator();
            while (cur.hasNext()) {
                SelectionKey oneKey = cur.next();
                cur.remove();
                try {
                    if (oneKey.isValid() && oneKey.isAcceptable()) {
                        Socket newSocket;
                        newSocket = socket.accept();
                        SocketChannel newChanel = newSocket.getChannel();
                        if (newChanel != null) {
                            newChanel.configureBlocking(false);
                            newChanel.register(selector, SelectionKey.OP_READ);
                        }
                    }
                    else {
                        if (oneKey.isValid() && oneKey.isReadable())
                            proxyOperations.readFromChanel(oneKey);
                        else {
                            if (oneKey.isValid() && oneKey.isWritable()) {
                                Related related = (Related) oneKey.attachment();
                                if (related.getCountMes() == 2)
                                    proxyOperations.thirdMessage(oneKey, related);
                                else
                                    proxyOperations.write(oneKey);
                            } else {
                                if (oneKey.isValid() && oneKey.isConnectable())
                                    proxyOperations.connect(oneKey);
                            }
                        }
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                    oneKey.interestOps(0);
                }
            }
        }
    }
}

File: src/main/kotlin/Proxy.kt
import java.net.DatagramSocket
import java.net.InetSocketAddress
import java.nio.channels.SelectionKey
import java.nio.channels.Selector
import java.nio.channels.ServerSocketChannel


 class Proxy  {
    private val proxyOperations = ProxyOperations()

     fun run(port : Int) {

        val socketChanel : ServerSocketChannel = ServerSocketChannel.open()
        socketChanel.configureBlocking(false)
        val socket = socketChanel.socket()
        socket.bind(InetSocketAddress(port))
        val selector = Selector.open()
        socketChanel.register(selector, SelectionKey.OP_ACCEPT)

        val dnsChannel = proxyOperations.createDatagramChannel();
         if(dnsChannel != null)
             dnsChannel.register(selector, SelectionKey.OP_READ);


        while(true) {
            val ready = selector.select()
            if(ready == 0) continue
            val keys = selector.selectedKeys()
            val cur = keys.iterator()
            while(cur.hasNext()) {
                val oneKey = cur.next() as SelectionKey
                cur.remove()
                if(!oneKey.isValid) {
                    continue
                }
                if(oneKey.isAcceptable) {
                    val newSocket = socket.accept()
                    val newChanel = newSocket.channel
                    if(newChanel != null) {
                        newChanel.configureBlocking(false)
                        newChanel.register(selector, SelectionKey.OP_READ)
                    }
                }
                if(oneKey.isReadable)
                    proxyOperations.readFromChanel(oneKey)
                if(!oneKey.isValid) {
                    continue
                }
                if(oneKey.isWritable) {
                    val related = oneKey.attachment() as Related
                    if(related.countMes == 2)
                        proxyOperations.thirdMessage(oneKey, related)
                    else
                        proxyOperations.write(oneKey)
                }
                if(!oneKey.isValid) {
                    continue
                }
                if(oneKey.isConnectable)
                    proxyOperations.connect(oneKey)
            }
        }
    }
}
